base/inference.jl

```diff
@@ -63,6 +63,13 @@ immutable Const
     Const(v::ANY) = new(v)
 end
 
+immutable PartialTypeVar
+    tv::TypeVar
+    lb_certain::Bool
+    ub_certain::Bool
```

Comment:
Yes, what @vtjnash said. You need to track identity through apply_type such that the UnionAll constructor works.